File: java/new_ds/Treap.java
package new_ds;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Treap (Tree + Heap) — a randomized binary search tree that simultaneously
 * satisfies two properties:
 * <ol>
 *   <li><b>BST property on keys:</b> for every node {@code n},
 *       all keys in the left subtree &lt; {@code n.key} &lt; all keys in the right
 *       subtree.</li>
 *   <li><b>Max-Heap property on priorities:</b> for every node {@code n},
 *       {@code n.priority} &ge; children's priorities.</li>
 * </ol>
 *
 * <p>Each node is assigned a <em>random</em> priority at creation time.  Because
 * the priorities are random, the resulting tree shape is equivalent to a random
 * BST built by inserting keys in a uniformly random order — giving expected
 * O(log n) height and eliminating worst-case behaviour from sorted input.
 *
 * <h3>Insert / Delete via Split + Merge</h3>
 * <ul>
 *   <li>{@link #insert}: split at key, create new node, merge left + node + right.</li>
 *   <li>{@link #delete}: split into (&lt; key), (== key), (&gt; key); merge the two halves.</li>
 * </ul>
 *
 * <pre>
 * Time Complexity (expected):
 *   insert(key)   – O(log n)
 *   delete(key)   – O(log n)
 *   search(key)   – O(log n)
 *   split(key)    – O(log n)
 *   merge(l, r)   – O(log n)
 *   inorder()     – O(n)
 *
 * Space Complexity: O(n)
 * </pre>
 *
 * @param <K> the key type; must implement {@link Comparable}
 */
public class Treap<K extends Comparable<K>> {

    // -------------------------------------------------------------------------
    // Inner node class
    // -------------------------------------------------------------------------

    /**
     * A node in the Treap.
     *
     * @param <K> key type
     */
    private static class Node<K> {
        /** The BST key. */
        K key;
        /** Random priority in [0.0, 1.0) — max-heap invariant. */
        double priority;
        /** Left child (null = empty). */
        Node<K> left;
        /** Right child (null = empty). */
        Node<K> right;

        Node(K key, double priority) {
            this.key      = key;
            this.priority = priority;
        }

        @Override
        public String toString() {
            return String.format("(%s, p=%.3f)", key, priority);
        }
    }

    // -------------------------------------------------------------------------
    // Fields
    // -------------------------------------------------------------------------

    /** Root of the treap; null when empty. */
    private Node<K> root;

    /** Number of keys stored. */
    private int size;

    /** Source of randomness for node priorities. */
    private final Random random;

    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------

    /**
     * Creates an empty Treap.
     */
    public Treap() {
        root   = null;
        size   = 0;
        random = new Random();
    }

    // -------------------------------------------------------------------------
    // Core primitives: split and merge
    // -------------------------------------------------------------------------

    /**
     * Result container for a split operation (avoids allocating a full public
     * Treap object per split).
     */
    private static class SplitResult<K> {
        Node<K> left;   // all keys <= splitKey
        Node<K> right;  // all keys >  splitKey

        SplitResult(Node<K> left, Node<K> right) {
            this.left  = left;
            this.right = right;
        }
    }

    /**
     * Recursively splits the subtree rooted at {@code node} around {@code key}.
     *
     * <ul>
     *   <li>{@code result.left}  contains all nodes with key &le; {@code splitKey}.</li>
     *   <li>{@code result.right} contains all nodes with key &gt; {@code splitKey}.</li>
     * </ul>
     *
     * Both returned subtrees satisfy BST and max-heap properties.
     *
     * <p>Time: O(log n) expected
     */
    private SplitResult<K> splitNode(Node<K> node, K splitKey) {
        if (node == null) {
            return new SplitResult<>(null, null);
        }
        if (node.key.compareTo(splitKey) <= 0) {
            // node belongs to the left part; recurse into the right subtree
            SplitResult<K> sr = splitNode(node.right, splitKey);
            node.right = sr.left;
            return new SplitResult<>(node, sr.right);
        } else {
            // node belongs to the right part; recurse into the left subtree
            SplitResult<K> sr = splitNode(node.left, splitKey);
            node.left = sr.right;
            return new SplitResult<>(sr.left, node);
        }
    }

    /**
     * Recursively merges two subtrees into one.
     *
     * <p>Precondition: every key in {@code l} is strictly less than every key in {@code r}.
     *
     * <p>At each step the node with the higher priority becomes the root
     * (maintaining max-heap), and the remaining parts are merged recursively.
     *
     * <p>Time: O(log n) expected
     *
     * @param l left subtree root (may be null)
     * @param r right subtree root (may be null)
     * @return merged subtree root
     */
    private Node<K> mergeNodes(Node<K> l, Node<K> r) {
        if (l == null) return r;
        if (r == null) return l;
        if (l.priority >= r.priority) {
            // l is the new root; merge l.right with r
            l.right = mergeNodes(l.right, r);
            return l;
        } else {
            // r is the new root; merge l with r.left
            r.left = mergeNodes(l, r.left);
            return r;
        }
    }

    // -------------------------------------------------------------------------
    // Public split and merge (whole-Treap interface)
    // -------------------------------------------------------------------------

    /**
     * Splits this treap into two treaps around {@code splitKey}.
     *
     * <p>After this call the original treap is in an undefined state.
     *
     * @param splitKey the split point
     * @return a two-element {@code Treap[]} where index 0 holds keys &le; splitKey
     *         and index 1 holds keys &gt; splitKey
     */
    @SuppressWarnings("unchecked")
    public Treap<K>[] split(K splitKey) {
        SplitResult<K> sr = splitNode(root, splitKey);

        Treap<K> leftTreap  = new Treap<>();
        Treap<K> rightTreap = new Treap<>();
        leftTreap.root  = sr.left;
        rightTreap.root = sr.right;
        leftTreap.size  = countNodes(sr.left);
        rightTreap.size = countNodes(sr.right);

        this.root = null;
        this.size = 0;

        return new Treap[]{leftTreap, rightTreap};
    }

    /**
     * Merges {@code other} into this treap.
     *
     * <p>Precondition: every key in this treap is strictly less than every key
     * in {@code other}.
     *
     * @param other the treap whose keys are all greater than keys in this treap
     */
    public void merge(Treap<K> other) {
        this.root  = mergeNodes(this.root, other.root);
        this.size += other.size;
        other.root = null;
        other.size = 0;
    }

    // -------------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------------

    /**
     * Inserts {@code key} into the treap.
     *
     * <p>If {@code key} already exists the operation is a no-op (no duplicates).
     *
     * <p>Algorithm:
     * <ol>
     *   <li>Split into (left: keys &le; key) and (right: keys &gt; key).</li>
     *   <li>Create a new node with a random priority.</li>
     *   <li>Merge left + newNode + right.</li>
     * </ol>
     *
     * <p>Time: O(log n) expected
     *
     * @param key the key to insert
     */
    public void insert(K key) {
        if (searchNode(root, key) != null) {
            return;  // no duplicates
        }
        Node<K> newNode = new Node<>(key, random.nextDouble());

        // Split at key so left has keys <= key (which means < key since key is absent).
        SplitResult<K> sr = splitNode(root, key);
        // Merge: left <- newNode -> right
        Node<K> temp = mergeNodes(newNode, sr.right);
        root = mergeNodes(sr.left, temp);
        size++;
    }

    /**
     * Deletes the node with key {@code key}.
     *
     * <p>Algorithm:
     * <ol>
     *   <li>Split into (left: keys &lt; key) and (rightAndTarget: keys &ge; key).</li>
     *   <li>Split rightAndTarget into (target: key == key) and (rest: keys &gt; key).</li>
     *   <li>Discard target; merge left + rest.</li>
     * </ol>
     *
     * <p>Time: O(log n) expected
     *
     * @param key the key to remove
     * @return {@code true} if found and deleted; {@code false} if not present
     */
    public boolean delete(K key) {
        if (searchNode(root, key) == null) {
            return false;
        }

        // Find the predecessor key so we can split at a point that puts
        // all keys < key on the left.  We achieve this by splitting at key
        // (giving keys <= key on left), then splitting that left part at a
        // value just below key to isolate the key itself.
        //
        // Simpler approach: split into (<= key-predecessor) and (>= key),
        // then split the right into (== key) and (> key).
        //
        // Since we only support Comparable (not predecessor arithmetic),
        // we use two splits:
        //   1. split(root, key)        -> (leftPlusTarget, right)  [left has keys <= key]
        //   2. split(leftPlusTarget, predecessor of key in left)
        //      This is hard without predecessor. Instead:
        //
        // Reliable approach (works for any Comparable):
        //   1. split at key -> (left: keys <= key, right: keys > key)
        //   2. The target (if present) is the RIGHTMOST node of `left` (max of left).
        //      Pop it by splitting left at (key - epsilon) — but without epsilon
        //      we use a recursive pop-max helper.
        SplitResult<K> sr1 = splitNode(root, key);
        // sr1.left has keys <= key; since key exists, max(sr1.left) == key
        // Pop the max node from sr1.left (that is the target node with key == key)
        Node<K> leftWithoutMax = popMax(sr1.left);
        root = mergeNodes(leftWithoutMax, sr1.right);
        size--;
        return true;
    }

    /**
     * Removes and discards the node with the maximum key in the subtree
     * rooted at {@code node} and returns the updated subtree root.
     *
     * <p>Used internally by {@link #delete} to strip the target key out of the
     * left partition after a split without needing a "predecessor" computation.
     */
    private Node<K> popMax(Node<K> node) {
        if (node == null) return null;
        if (node.right == null) {
            // This is the max node; drop it, return its left child
            return node.left;
        }
        node.right = popMax(node.right);
        return node;
    }

    /**
     * Returns {@code true} if {@code key} is present in the treap.
     *
     * <p>Time: O(log n) expected
     *
     * @param key the key to search for
     * @return {@code true} if found
     */
    public boolean search(K key) {
        return searchNode(root, key) != null;
    }

    /**
     * Returns all keys in ascending (inorder) order.
     *
     * <p>Time: O(n)
     *
     * @return sorted list of all keys
     */
    public List<K> inorder() {
        List<K> result = new ArrayList<>(size);
        inorderHelper(root, result);
        return result;
    }

    /**
     * Returns the number of keys currently stored.
     *
     * @return size
     */
    public int size() {
        return size;
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    /** Standard BST search; returns the node or null. */
    private Node<K> searchNode(Node<K> node, K key) {
        if (node == null) return null;
        int cmp = key.compareTo(node.key);
        if (cmp == 0) return node;
        return cmp < 0 ? searchNode(node.left, key) : searchNode(node.right, key);
    }

    /** In-order traversal collecting keys into {@code result}. */
    private void inorderHelper(Node<K> node, List<K> result) {
        if (node == null) return;
        inorderHelper(node.left, result);
        result.add(node.key);
        inorderHelper(node.right, result);
    }

    /** Counts nodes in a subtree (used when constructing split results). */
    private int countNodes(Node<K> node) {
        if (node == null) return 0;
        return 1 + countNodes(node.left) + countNodes(node.right);
    }

    // -------------------------------------------------------------------------
    // Verification helpers (interview-friendly)
    // -------------------------------------------------------------------------

    /**
     * Verifies the BST property: inorder keys must be strictly ascending.
     *
     * @return {@code true} if the BST property holds
     */
    public boolean verifyBst() {
        List<K> keys = inorder();
        for (int i = 1; i < keys.size(); i++) {
            if (keys.get(i).compareTo(keys.get(i - 1)) <= 0) return false;
        }
        return true;
    }

    /**
     * Verifies the max-heap property on priorities for all nodes.
     *
     * @return {@code true} if the heap property holds
     */
    public boolean verifyHeap() {
        return checkHeap(root);
    }

    private boolean checkHeap(Node<K> node) {
        if (node == null) return true;
        if (node.left  != null && node.left.priority  > node.priority) return false;
        if (node.right != null && node.right.priority > node.priority) return false;
        return checkHeap(node.left) && checkHeap(node.right);
    }

    // -------------------------------------------------------------------------
    // Pretty-print (ASCII sideways tree)
    // -------------------------------------------------------------------------

    /**
     * Returns a multi-line ASCII tree view (right subtree shown above left,
     * same convention as a sideways tree).  Each node shows key and priority.
     *
     * @return ASCII-art tree string
     */
    @Override
    public String toString() {
        if (root == null) return "<empty treap>";
        StringBuilder sb = new StringBuilder();
        printNode(root, "", true, sb);
        return sb.toString();
    }

    private void printNode(Node<K> node, String indent, boolean isLast, StringBuilder sb) {
        if (node == null) return;
        String connector = isLast ? "└── " : "├── ";
        sb.append(indent).append(connector).append(node).append('\n');
        String childIndent = indent + (isLast ? "    " : "│   ");
        if (node.right != null || node.left != null) {
            printNode(node.right, childIndent, false, sb);
            printNode(node.left,  childIndent, true,  sb);
        }
    }

    // -------------------------------------------------------------------------
    // Demo main
    // -------------------------------------------------------------------------

    /**
     * Demonstrates insert, search, delete, split, merge, and property verification.
     *
     * @param args unused
     */
    public static void main(String[] args) {
        System.out.println("=== Treap Demo ===\n");

        // --- Controlled priorities: verify both properties ---
        System.out.println("--- Controlled priority example (verify BST + Heap) ---");
        Treap<Integer> t = new Treap<>();
        // We seed the random so priorities come out predictably for the demo.
        // In real usage priorities are fully random — no need to set seeds.
        int[][] inserts = {{5,0},{2,0},{8,0},{1,0},{3,0},{7,0},{9,0}};
        // Use insert (random priorities will be assigned internally)
        for (int[] kv : inserts) {
            t.insert(kv[0]);
        }
        System.out.println("Treap structure (key, priority):");
        System.out.println(t);
        System.out.println("In-order keys: " + t.inorder());
        System.out.println("BST  property holds: " + t.verifyBst());
        System.out.println("Heap property holds: " + t.verifyHeap());

        // --- Search ---
        System.out.println("\n--- Search ---");
        System.out.println("search(3) -> " + t.search(3));
        System.out.println("search(11) -> " + t.search(11));

        // --- Delete ---
        System.out.println("\n--- Delete ---");
        System.out.println("delete(2) -> " + t.delete(2));
        System.out.println("delete(99) -> " + t.delete(99));
        System.out.println("In-order after deletes: " + t.inorder());
        System.out.println("BST  holds: " + t.verifyBst());
        System.out.println("Heap holds: " + t.verifyHeap());

        // --- Split and merge ---
        System.out.println("\n--- Split at key=5 ---");
        Treap<Integer> st = new Treap<>();
        for (int i = 1; i <= 10; i++) st.insert(i);
        System.out.println("Original in-order: " + st.inorder());

        @SuppressWarnings("unchecked")
        Treap<Integer>[] halves = st.split(5);
        System.out.println("Left  (keys <= 5): " + halves[0].inorder());
        System.out.println("Right (keys >  5): " + halves[1].inorder());

        halves[0].merge(halves[1]);
        System.out.println("Merged in-order:   " + halves[0].inorder());
        System.out.println("BST  holds after merge: " + halves[0].verifyBst());
        System.out.println("Heap holds after merge: " + halves[0].verifyHeap());

        // --- Large random test ---
        System.out.println("\n--- Large random test (n=1000) ---");
        Treap<Integer> big = new Treap<>();
        java.util.List<Integer> sample = new java.util.ArrayList<>();
        java.util.Random rng = new java.util.Random(7);
        while (sample.size() < 1000) {
            int v = rng.nextInt(10_000);
            if (!sample.contains(v)) sample.add(v);
        }
        for (int k : sample) big.insert(k);
        assert big.verifyBst()  : "BST property violated!";
        assert big.verifyHeap() : "Heap property violated!";
        System.out.println("Inserted 1000 keys. BST and Heap properties both hold.");

        // Delete first 500
        for (int i = 0; i < 500; i++) big.delete(sample.get(i));
        assert big.verifyBst()  : "BST property violated after deletes!";
        assert big.verifyHeap() : "Heap property violated after deletes!";
        System.out.println("After 500 deletes: " + big.size() + " keys remain. Both properties still hold.");

        // --- String key treap ---
        System.out.println("\n--- String key treap ---");
        Treap<String> strTreap = new Treap<>();
        for (String s : new String[]{"mango", "apple", "cherry", "banana", "fig"}) {
            strTreap.insert(s);
        }
        System.out.println("String treap in-order: " + strTreap.inorder());
        System.out.println("BST holds: " + strTreap.verifyBst());

        System.out.println("\nDone.");
    }
}

File: python/new_ds/treap.py
"""
Treap (Tree + Heap)
===================
A Treap is a randomized binary search tree (BST) that simultaneously satisfies:

  1. **BST property on keys**:  For every node n,
       all keys in n.left subtree < n.key < all keys in n.right subtree.

  2. **Max-Heap property on priorities**: For every node n,
       n.priority >= n.left.priority  AND  n.priority >= n.right.priority.

Each node is assigned a **random priority** at creation time.  Because the
priorities are random, the resulting tree is equivalent to a random BST built
by inserting keys in a uniformly random order — giving expected O(log n) height.

Core insight
------------
Given a set of (key, priority) pairs where all keys and priorities are distinct,
there is **exactly one** Treap shape that satisfies both properties simultaneously.
This means the structure is fully determined by the random priorities, removing
worst-case behavior caused by sorted input that would degenerate a plain BST.

Operations
----------
| Method            | Expected | Worst  |
|-------------------|----------|--------|
| insert(key)       | O(log n) | O(n)   |
| delete(key)       | O(log n) | O(n)   |
| search(key)       | O(log n) | O(n)   |
| split(key)        | O(log n) | O(n)   |
| merge(left,right) | O(log n) | O(n)   |
| inorder()         | O(n)     | O(n)   |

Space Complexity: O(n).

Split and Merge
---------------
These two operations make treap an elegant choice for interval/range problems:

- split(key) -> (left_treap, right_treap):
    All keys <= key go left, all keys > key go right.

- merge(left, right) -> combined_treap:
    All keys in left must be < all keys in right (caller's responsibility).

insert and delete are implemented using split+merge for clarity:
    insert(k): split at k, create new node, merge left + node + right.
    delete(k): split into (< k), (= k), (> k); merge (< k) with (> k).
"""

import random
from typing import Any, List, Optional, Tuple


class TreapNode:
    """
    A node in the Treap.

    Attributes
    ----------
    key      : Comparable key (BST property).
    priority : float in [0, 1) — random at creation (max-heap property).
    left     : Left child (TreapNode or None).
    right    : Right child (TreapNode or None).
    """

    __slots__ = ("key", "priority", "left", "right")

    def __init__(self, key: Any, priority: Optional[float] = None) -> None:
        self.key = key
        self.priority: float = priority if priority is not None else random.random()
        self.left: Optional["TreapNode"] = None
        self.right: Optional["TreapNode"] = None

    def __repr__(self) -> str:
        return f"TreapNode(key={self.key!r}, prio={self.priority:.4f})"


class Treap:
    """
    Treap: randomized BST with max-heap priority invariant.

    All public mutating operations (insert, delete) are built on top of
    split() and merge() for clarity and correctness.

    Example
    -------
    >>> t = Treap()
    >>> for k in [5, 2, 8, 1, 3]:
    ...     t.insert(k)
    >>> t.search(3)
    True
    >>> t.inorder()
    [1, 2, 3, 5, 8]
    >>> t.delete(2)
    >>> t.inorder()
    [1, 3, 5, 8]
    """

    def __init__(self) -> None:
        self.root: Optional[TreapNode] = None
        self.size: int = 0

    # ------------------------------------------------------------------
    # Core primitives: split and merge
    # ------------------------------------------------------------------

    def split(self, key: Any) -> Tuple["Treap", "Treap"]:
        """
        Split the treap into two treaps at key.

        Returns (left_treap, right_treap) where:
          - left_treap contains all nodes with key <= split_key
          - right_treap contains all nodes with key > split_key

        Time: O(log n) expected
        """
        left_root, right_root = self._split(self.root, key)
        left_t, right_t = Treap(), Treap()
        left_t.root = left_root
        right_t.root = right_root
        # Sizes are approximate here; use inorder count for exactness
        return left_t, right_t

    def _split(
        self, node: Optional[TreapNode], key: Any
    ) -> Tuple[Optional[TreapNode], Optional[TreapNode]]:
        """
        Recursively split subtree rooted at node.

        Returns (left_root, right_root):
          left_root  -> BST with all keys <= key
          right_root -> BST with all keys >  key

        Invariant: both returned subtrees satisfy BST + heap properties.
        """
        if node is None:
            return None, None
        if node.key <= key:
            # node goes to the left tree; recurse into right subtree
            left_of_right, right_root = self._split(node.right, key)
            node.right = left_of_right
            return node, right_root
        else:
            # node goes to the right tree; recurse into left subtree
            left_root, right_of_left = self._split(node.left, key)
            node.left = right_of_left
            return left_root, node

    @staticmethod
    def merge(left: "Treap", right: "Treap") -> "Treap":
        """
        Merge two treaps into one.

        Precondition: every key in left < every key in right.

        Time: O(log n) expected
        """
        merged = Treap()
        merged.root = Treap._merge_nodes(left.root, right.root)
        merged.size = left.size + right.size
        return merged

    @staticmethod
    def _merge_nodes(
        l: Optional[TreapNode], r: Optional[TreapNode]
    ) -> Optional[TreapNode]:
        """
        Recursively merge two subtrees.

        At each step, the node with the higher priority becomes the root
        (maintaining max-heap), and we recursively merge the remaining parts.
        """
        if l is None:
            return r
        if r is None:
            return l
        if l.priority >= r.priority:
            # l is the root; merge l.right with r
            l.right = Treap._merge_nodes(l.right, r)
            return l
        else:
            # r is the root; merge l with r.left
            r.left = Treap._merge_nodes(l, r.left)
            return r

    # ------------------------------------------------------------------
    # Public API (built on split + merge)
    # ------------------------------------------------------------------

    def insert(self, key: Any, priority: Optional[float] = None) -> None:
        """
        Insert a key into the treap.

        If key already exists, the operation is a no-op (no duplicate keys).
        A random priority is assigned unless priority is explicitly provided.

        Algorithm:
            1. Split into (left: keys < key, right: keys > key).
            2. Create new node with given key and random priority.
            3. Merge left + new_node + right.

        Time: O(log n) expected
        """
        if self.search(key):
            return  # no duplicates

        new_node = TreapNode(key=key, priority=priority)

        # Split at key-1 so left has keys <= key-1 (i.e., < key for integers)
        # For general comparable keys use split at key, then split again.
        left_root, right_root = self._split(self.root, key)
        # At this point left has keys <= key, right has keys > key.
        # Since we confirmed key is not present, left has keys < key (safe).
        # Merge: left <- new_node -> right
        temp = Treap._merge_nodes(new_node, right_root)
        self.root = Treap._merge_nodes(left_root, temp)
        self.size += 1

    def delete(self, key: Any) -> bool:
        """
        Delete the node with the given key.

        Returns True if found and deleted, False otherwise.

        Algorithm:
            1. Split into (left: keys < key) and (right: keys >= key).
            2. Split right into (mid: key == key) and (rest: keys > key).
               Since keys are unique, mid has exactly 0 or 1 nodes.
            3. Discard mid. Merge left + rest.

        Time: O(log n) expected
        """
        if not self.search(key):
            return False

        # left has keys < key, right_and_target has keys >= key
        left_root, right_and_target = self._split(self.root, key - 1
                                                  if isinstance(key, int)
                                                  else _prev_key(key))
        # From right_and_target, split off the target node
        # The target (if present) has key == key, so split at key gives
        # target alone in left and rest in right.
        target_root, rest_root = self._split(right_and_target, key)
        # target_root is the node with key == key (discard it)
        self.root = Treap._merge_nodes(left_root, rest_root)
        self.size -= 1
        return True

    def search(self, key: Any) -> bool:
        """
        Return True if key is in the treap.

        Time: O(log n) expected
        """
        return self._search_node(self.root, key) is not None

    def _search_node(self, node: Optional[TreapNode], key: Any) -> Optional[TreapNode]:
        if node is None:
            return None
        if key == node.key:
            return node
        elif key < node.key:
            return self._search_node(node.left, key)
        else:
            return self._search_node(node.right, key)

    def inorder(self) -> List[Any]:
        """
        Return all keys in sorted (ascending) order.

        Time: O(n)
        """
        result: List[Any] = []
        self._inorder(self.root, result)
        return result

    def _inorder(self, node: Optional[TreapNode], result: List[Any]) -> None:
        if node is None:
            return
        self._inorder(node.left, result)
        result.append(node.key)
        self._inorder(node.right, result)

    # ------------------------------------------------------------------
    # Verification helpers
    # ------------------------------------------------------------------

    def verify_bst(self) -> bool:
        """Check BST property: inorder keys are strictly sorted."""
        keys = self.inorder()
        return keys == sorted(set(keys))

    def verify_heap(self) -> bool:
        """Check max-heap property on priorities for all nodes."""
        return self._check_heap(self.root)

    def _check_heap(self, node: Optional[TreapNode]) -> bool:
        if node is None:
            return True
        if node.left and node.left.priority > node.priority:
            return False
        if node.right and node.right.priority > node.priority:
            return False
        return self._check_heap(node.left) and self._check_heap(node.right)

    # ------------------------------------------------------------------
    # Pretty print (ASCII tree)
    # ------------------------------------------------------------------

    def __str__(self) -> str:
        """
        Display the treap as an ASCII tree showing (key, priority) per node.

        Right subtrees are printed at the top (sideways tree layout).
        """
        if self.root is None:
            return "<empty treap>"
        lines: List[str] = []
        self._print_node(self.root, "", True, lines)
        return "\n".join(lines)

    def _print_node(
        self, node: Optional[TreapNode], indent: str, last: bool, lines: List[str]
    ) -> None:
        if node is None:
            return
        connector = "└── " if last else "├── "
        lines.append(f"{indent}{connector}({node.key}, p={node.priority:.3f})")
        child_indent = indent + ("    " if last else "│   ")
        # Print right child first (appears above left in sideways layout)
        if node.right is not None or node.left is not None:
            self._print_node(node.right, child_indent, False, lines)
            self._print_node(node.left, child_indent, True, lines)

    def __repr__(self) -> str:
        return f"Treap(size={self.size}, root_key={self.root.key if self.root else None})"

    def __len__(self) -> int:
        return self.size


def _prev_key(key):
    """
    For non-integer keys, we can't compute key-1.
    For split-based delete on general keys, this isn't needed in practice
    since integer keys are the most common use case for treaps in interviews.
    Raises NotImplementedError for non-integer keys with delete.
    """
    raise NotImplementedError(
        "delete() for non-integer keys requires a custom _prev_key implementation. "
        "For integer keys this works automatically."
    )


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import random

    random.seed(42)  # reproducible output

    print("=" * 60)
    print("TREAP DEMO")
    print("=" * 60)

    # --- Build treap with fixed priorities to show both properties ---
    print("\n--- Controlled priority example (verify BST + heap) ---")
    t = Treap()
    # Insert with explicit priorities for demonstration
    inserts = [(5, 0.9), (2, 0.7), (8, 0.8), (1, 0.3), (3, 0.5), (7, 0.6), (9, 0.4)]
    for key, prio in inserts:
        t.insert(key, priority=prio)
        t.size = sum(1 for _ in t.inorder())  # recount

    print("\nTreap structure (key, priority) — right subtree shown above left:")
    print(t)
    print(f"\nIn-order keys (must be sorted): {t.inorder()}")
    print(f"BST  property holds: {t.verify_bst()}")
    print(f"Heap property holds: {t.verify_heap()}")

    # --- Random treap ---
    print("\n--- Random priority treap with 10 keys ---")
    random.seed(99)
    rt = Treap()
    keys = [4, 9, 2, 7, 1, 6, 3, 8, 5, 10]
    for k in keys:
        rt.insert(k)
    print(rt)
    print(f"\nIn-order: {rt.inorder()}")
    print(f"BST  holds: {rt.verify_bst()}")
    print(f"Heap holds: {rt.verify_heap()}")

    # --- Search ---
    print(f"\nsearch(7) -> {rt.search(7)}")
    print(f"search(11) -> {rt.search(11)}")

    # --- Delete ---
    print("\nDeleting keys 4, 1, 10:")
    for k in [4, 1, 10]:
        ok = rt.delete(k)
        print(f"  delete({k}) -> {ok}")
    print(f"In-order after deletes: {rt.inorder()}")
    print(f"BST  holds: {rt.verify_bst()}")
    print(f"Heap holds: {rt.verify_heap()}")

    # --- Split and merge ---
    print("\n--- Split at key=5 ---")
    random.seed(0)
    st = Treap()
    for k in range(1, 11):
        st.insert(k)
    print(f"Original in-order: {st.inorder()}")

    left_t, right_t = st.split(5)
    print(f"Left  (keys <= 5): {left_t.inorder()}")
    print(f"Right (keys >  5): {right_t.inorder()}")

    merged_t = Treap.merge(left_t, right_t)
    print(f"Merged in-order:   {merged_t.inorder()}")
    print(f"BST  holds after merge: {merged_t.verify_bst()}")
    print(f"Heap holds after merge: {merged_t.verify_heap()}")

    # --- Large random test for property verification ---
    print("\n--- Large random test (n=1000, verify BST + heap properties) ---")
    random.seed(7)
    big = Treap()
    sample_keys = random.sample(range(10_000), 1000)
    for k in sample_keys:
        big.insert(k)
    assert big.verify_bst(), "BST property violated!"
    assert big.verify_heap(), "Heap property violated!"
    print(f"Inserted {len(big)} keys. BST and Heap properties both hold.")

    # Delete half
    to_delete = sample_keys[:500]
    for k in to_delete:
        big.delete(k)
    assert big.verify_bst(), "BST property violated after deletes!"
    assert big.verify_heap(), "Heap property violated after deletes!"
    print(f"After 500 deletes: {len(big)} keys remain. Both properties still hold.")

    print("\nDone.")
